p2p/transport/tcp/tcp.go

```diff
@@ -0,0 +1,418 @@
+}
+
+// MultiplexTransportMaxIncomingConnections sets the maximum number of
+// simultaneous connections (incoming). Default: 0 (unlimited).
+func MultiplexTransportMaxIncomingConnections(n int) MultiplexTransportOption {
+	return func(mt *MultiplexTransport) { mt.maxIncomingConnections = n }
+}
+
+// MultiplexTransport accepts and dials tcp connections and upgrades them to
+// multiplexed peers.
+type MultiplexTransport struct {
+	netAddr                na.NetAddress
+	listener               net.Listener
+	maxIncomingConnections int // see MaxIncomingConnections
+
+	acceptc chan accept
+	closec  chan struct{}
+
+	// Lookup table for duplicate ip and id checks.
+	conns       ConnSet
+	connFilters []ConnFilterFunc
+
+	dialTimeout      time.Duration
+	filterTimeout    time.Duration
+	handshakeTimeout time.Duration
+	nodeKey          nodekey.NodeKey
+	resolver         IPResolver
+
+	// TODO(xla): This config is still needed as we parameterise peerConn and
+	// peer currently. All relevant configuration should be refactored into options
+	// with sane defaults.
+	mConfig conn.MConnConfig
+}
+
+// Test multiplexTransport for interface completeness.
+var (
+	_ transportLifecycle = (*MultiplexTransport)(nil)
+)
+
+// NewMultiplexTransport returns a tcp connected multiplexed peer.
+func NewMultiplexTransport(
+	nodeKey nodekey.NodeKey,
+	mConfig conn.MConnConfig,
+) *MultiplexTransport {
+	return &MultiplexTransport{
+		acceptc:          make(chan accept),
+		closec:           make(chan struct{}),
+		dialTimeout:      defaultDialTimeout,
+		filterTimeout:    defaultFilterTimeout,
+		handshakeTimeout: defaultHandshakeTimeout,
+		mConfig:          mConfig,
+		nodeKey:          nodeKey,
+		conns:            NewConnSet(),
+		resolver:         net.DefaultResolver,
+	}
+}
+
+// NetAddress implements Transport.
+func (mt *MultiplexTransport) NetAddress() na.NetAddress {
+	return mt.netAddr
+}
+
+// Accept implements Transport.
+func (mt *MultiplexTransport) Accept() (net.Conn, *na.NetAddress, error) {
+	select {
+	// This case should never have any side-effectful/blocking operations to
+	// ensure that quality peers are ready to be used.
+	case a := <-mt.acceptc:
+		if a.err != nil {
+			return nil, nil, a.err
+		}
+
+		// cfg.outbound = false
+
+		return a.conn, a.netAddr, nil
+	case <-mt.closec:
+		return nil, nil, ErrTransportClosed{}
+	}
+}
+
+// Dial implements Transport.
+func (mt *MultiplexTransport) Dial(
+	addr na.NetAddress,
+) (net.Conn, error) {
+	c, err := addr.DialTimeout(mt.dialTimeout)
+	if err != nil {
+		return nil, err
+	}
+
+	if mt.mConfig.TestFuzz {
+		// so we have time to do peer handshakes and get set up.
+		c = fuzz.ConnAfterFromConfig(c, 10*time.Second, mt.mConfig.TestFuzzConfig)
+	}
+
+	// TODO(xla): Evaluate if we should apply filters if we explicitly dial.
+	if err := mt.filterConn(c); err != nil {
+		return nil, err
+	}
+
+	secretConn, err := mt.upgrade(c, &addr)
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
just noticed that in the previous file, there was a cfg parameter and the `outbound` was set. I'm assuming this is not needed or was removed on purpose but just wanted to highlight here
```
cfg.outbound = true
```